File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Declare and initialize a variable
    long long cardNumber = 0;
    
    // Prompt user input
    do
    {
        cardNumber = get_long_long("Enter your credit card number: \n");
    }
    while (cardNumber < 0);

    // Determine the validity of card by couting the number of digits
    int count = 0;
    long long digits = cardNumber;
    while (digits > 0)
    {
        // Repeatedly divide cardNumber by 10 to remove the last digit
        digits = digits / 10;
        //Increment the count by each digit
        count++;
    }

    // Rule out irrelevant number of digits
    if (count != 13 && count != 15 && count != 16)
    {
        printf("INVALID\n");
    }

    // Delcare a "number" array to store digit counts of card numbers
    int number[count];
    for (int i = 0; i < count; i++)
    {
        number[i] = cardNumber % 10;
        cardNumber = cardNumber / 10;
    }

    // Declare an array for the orginal card number using the above "number" array
    int originalNumber[count];
    for (int i = 1; i < count; i++)
    {
        originalNumber[i] = number[i];
    }

    // Multiply digits by starting with the number’s second-to-last digit
    for (int i = 1; i < count; i += 2)
    {
        number[i] = number[i] * 2;
    }

    int v = 0;
    int temp;
    
    // Validating 13-digit VISA cards
    if (count == 13)
    {
        for (int i = 0; i < count; i++)
        {
            // Add the sum of products' digits to the sum of the digits that weren't multiplied
            temp = (number[i] % 10) + (number[i] / 10 % 10);
            v = v + temp;
        }
        if (originalNumber[12] == 12 && v % 10 == 0)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // Validating AMEX cards
    if (count == 15)
    {
        for (int i = 0; i < count; i++)
        {
            temp = (number[i] % 10) + (number[i] / 10 % 10);
            v = v+ temp;
        }

        // Condition for AMEX cards: card number starting with 34 or 37)
        if (originalNumber[14] == 3 && v % 10 == 0 && (originalNumber[13] == 4 || 
            originalNumber[13]  == 7))
        {
            printf("AMEX\n");
        }
        
        else
        {
            printf("INVALID\n");
        }
    }

    // 16-digit VISA cards & MASTERCARD
    if (count == 16)
    {
        for (int i = 0; i < count; i++)
        {
            temp = (number[i] % 10) + (number[i] / 10 % 10);
            v = v + temp;
        }

        if (originalNumber[15] == 4 && v % 10 == 0)
        {
            printf("VISA\n");
        }

        else if (originalNumber[15] == 5 && v % 10 == 0 && (originalNumber[14] == 1 || 
                 originalNumber[14] == 2 || 
                 originalNumber[14] == 3 || 
                 originalNumber[14] == 4 ||                   
                 originalNumber[14] == 5))
        {
            printf("MASTERCARD\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }
}

File: credit.py
while True:
    number = int(input("Please enter your card number: "))
    if number > 0:
        break

# Store digits in a list
digits = list(str(number))

# If there is an invalid number of digits
if len(digits) != 13 and len(digits) != 15 and len(digits) != 16:
    print("Invalid number")

# Take every 2-nth element of the number into a new list
underlined = digits[::2]

# Multiply each of the underlined digit by 2
product = [i * 2 for i in map(int, underlined)]

# Join all the digits for splitting later
product = int(''.join(map(str, product)))

# Sum those products’ digit
prod_digits = [int(i) for i in str(product)]    # Convert string digits in list to type integer
sum_product = sum(prod_digits)

# Sum of un-multiplied digits from line 7
un_m = sum([int(i) for i in digits[1::2]])      # Convert string digits in sliced list to type integer then sum the integers

# Sum of products' sum and un-multiplied sum
valid = sum_product + un_m
valid = list(str(valid))

# Validating VISA
if (len(digits) == 13 or len(digits) == 16) and valid[1] == "0" and digits[0] == "4":
    print("VISA\n")

# Validating MASTERCARD
elif len(digits) == 16 and valid[1] =="0" and ( digits[:2] == ["5", "1"] or
                                                digits[:2] == ["5", "2"] or
                                                digits[:2] == ["5", "3"] or
                                                digits[:2] == ["5", "4"] or
                                                digits[:2] == ["5", "5"]):
    print("MASTERCARD\n")

# Validating AMEX
elif len(digits) == 15 and (digits[:2] == ["3", "4"] or digits[:2] == ["3", "7"]):
    print("AMEX\n")

else:
    print("INVALID\n")
